env/interpolator.go

```diff
@@ -2,156 +2,162 @@ package env
 
-func Interpolate(str string) (interpolated string, err error) {
+// Interpolate takes a string and interpolates shell-like variable expansions in the provided string
+func (e Environment) Interpolate(str string) (string, error) {
+	var err error
+
 	// Do a parse and handle ENV variables with the ${} syntax, i.e. ${FOO}
-	interpolated = variablesWithBracketsRegex.ReplaceAllStringFunc(str, func(v string) string {
-		if err == nil {
-			key, option := extractKeyAndOptionFromVariable(v)
+	interpolated := variablesWithBracketsRegex.ReplaceAllStringFunc(str, func(v string) string {
+		// if there has been an error previously, skip interpolation
+		if err != nil {
+			return v
+		}
+		key, option := extractKeyAndOptionFromVariable(v)
 
-			// Just return the key by itself if it was escaped
-			if isPrefixedWithEscapeSequence(v) {
-				v = key
-			} else {
-				err = isValidPosixEnvironmentVariable(v)
-				if err != nil {
-					return v
-				}
+		// Just return the key by itself if it was escaped
+		if isPrefixedWithEscapeSequence(v) {
+			return key
+		}
 
-				vv, isEnvironmentVariableSet := os.LookupEnv(key)
-
-				switch {
-				case substringRegexp.MatchString(option):
-					// Substring Expansion -- select a substring of a variable with:
-					//
-					// ${parameter:offset}
-					// ${parameter:offset:length}
-					//
-					// In the first form select a substring of $parameter starting from
-					// 0-indexed offset until the end of $parameter. If offset is
-					// negative then it is an offset from the end of $parameter instead.
-					//
-					// In the second form, length is the number of characters from offset
-					// to select. If negeative, length is instead an offset from the end
-					// of $parameter.
-					//
-					match := substringRegexp.FindStringSubmatch(option)
-					lenvv := int64(len(vv))
-
-					offset, err := strconv.ParseInt(match[1], 10, 0)
-					if err != nil {
-						fmt.Println(err)
-						return v
-					}
+		err = isValidPosixEnvironmentVariable(v)
+		if err != nil {
+			return v
+		}
 
-					// Negative offsets = from end
-					if offset < 0 {
-						offset = lenvv - (-offset)
-					}
+		vv := e.Get(key)
+		isEnvironmentVariableSet := e.Exists(key)
+
+		switch {
+		case substringRegexp.MatchString(option):
+			// Substring Expansion -- select a substring of a variable with:
+			//
+			// ${parameter:offset}
+			// ${parameter:offset:length}
+			//
+			// In the first form select a substring of $parameter starting from
+			// 0-indexed offset until the end of $parameter. If offset is
+			// negative then it is an offset from the end of $parameter instead.
+			//
+			// In the second form, length is the number of characters from offset
+			// to select. If negeative, length is instead an offset from the end
+			// of $parameter.
+			//
+			match := substringRegexp.FindStringSubmatch(option)
+			lenvv := int64(len(vv))
+
+			offset, parseErr := strconv.ParseInt(match[1], 10, 0)
+			if err != nil {
+				err = parseErr
```

Comment:
Actually this is a change, if the parseInt fails, it returns an error, before it just logged it